### PR TITLE
Fix for #412

### PR DESCRIPTION
**Pull request by empirephoenix**
Use the newer calculation for the i2c

### CODE REVIEW

**Comment by empirephoenix** [comment]
@ivmarkov I would need some advice, log2 is not available in std but would be required for the calculation, I could do this with ilog2, but then I have no idea how to ensure the ceiling() is done correctly. Any advice?

**Comment by ivmarkov** [comment]
> @ivmarkov I would need some advice, log2 is not available in std but would be required for the calculation, I could do this with ilog2, but then I have no idea how to ensure the ceiling() is done correctly. Any advice?

... assuming you do the math with `u64`, the following might work:
`abc.log2() + (if abc.leading_zeroes() + abc.trailing_zeroes() + 1 < 64 { 1 } else { 0 })`

NOTE 1: Ideally, you should avoid floating point math anyway, on MCUs and esps specifically.
NOTE 2: I have not checked yet your computation and why you need to take the logarithm of something

**Comment by empirephoenix** [comment]
Ok this seems to work, 104ms returns 22 for the register while 105 rounds up to 23 and is properly failing

File: src/i2c.rs [hunk 1]
```diff
@@ -19,16 +19,38 @@ crate::embedded_hal_error!(
...
         APBTickType(
             ((duration.as_nanos() + APB_TICK_PERIOD_NS as u128 - 1) / APB_TICK_PERIOD_NS as u128)
                 as ::core::ffi::c_int,
         )
     }
+    #[cfg(not(any(esp32, esp32s2)))]
+    /// Conversion for newer esp models, be aware, that the hardware can only represent 22 different values, values will be rounded to the next larger valid one. Calculation only valid for 40mhz clock source
+    fn from(duration: Duration) -> Self {
+        let target_ns = duration.as_nanos();
+        let target_ns_f = target_ns as u64;
```
Comment (ivmarkov) [inline src/i2c.rs:40]: This is no longer floating point so can you remove the `_f` suffix? Also, do you actually need this variable? Isn't `as_nanos` returning `u64` already?

File: src/i2c.rs [hunk 2]
```diff
@@ -19,16 +19,38 @@ crate::embedded_hal_error!(
...
+        let abc = target_ns_f / (XTAL_TICK_PERIOD_NS as u64);
```
Comment (ivmarkov) [inline src/i2c.rs:41]: What does `abc` stand for? Can you name it in a bit more descriptive way?

File: src/i2c.rs [hunk 3]
```diff
@@ -19,16 +19,38 @@ crate::embedded_hal_error!(
...
+        let i = abc.ilog2()
```
Comment (ivmarkov) [inline src/i2c.rs:42]: Ditto for `i`

File: src/i2c.rs [hunk 4]
```diff
@@ -19,16 +19,38 @@ crate::embedded_hal_error!(
...
+            + (if abc.leading_zeros() + abc.trailing_zeros() + 1 < 64 {
+                1
+            } else {
+                0
+            });
+        if i <= 22 {
```
Comment (ivmarkov) [inline src/i2c.rs:48]: Doesn't esp idf define a constant already for `22`? If yes, can you use it?

File: src/i2c.rs [hunk 5]
```diff
@@ -19,16 +19,38 @@ crate::embedded_hal_error!(
...
+            return APBTickType(i as ::core::ffi::c_int);
+        }
+        //produce an error in the lower set_i2c_timeout, so the user is informed that the requested timeout is larger than the next valid one.
+        APBTickType(32 as ::core::ffi::c_int)
```
Comment (ivmarkov) [inline src/i2c.rs:52]: What does 32 mean? Maybe we should just return the esp idf constant (see above) + 1?

File: src/i2c.rs [hunk 6]
```diff
@@ -19,16 +19,38 @@ crate::embedded_hal_error!(
     embedded_hal::i2c::ErrorKind
 );
 
+#[cfg(any(esp32, esp32s2))]
```
Comment (ivmarkov) [inline src/i2c.rs:22]: Aren't you losing precision here? 1 000 000 000 is not divisible by 80 000 000?

File: src/i2c.rs [hunk 7]
```diff
@@ -19,16 +19,38 @@ crate::embedded_hal_error!(
...
 const APB_TICK_PERIOD_NS: u32 = 1_000_000_000 / 80_000_000;
+#[cfg(not(any(esp32, esp32s2)))]
+const XTAL_TICK_PERIOD_NS: u32 = 1_000_000_000 / XTAL_CLK_FREQ;
```
Comment (ivmarkov) [inline src/i2c.rs:25]: Here I'm not sure, but you should check.

**Review by ivmarkov — CHANGES_REQUESTED** [review]

File: src/i2c.rs (hunk 7 above)
Comment (empirephoenix) [reply src/i2c.rs:25]: its 25ns per clock actually

File: src/i2c.rs (hunk 6 above)
Comment (empirephoenix) [reply src/i2c.rs:22]: That one was already used before, I only added the XTAL based 40mhz calculation, but you are right, it is 12.5ns so the old calculation might be wrong by a few clocks

File: src/i2c.rs (hunk 5 above)
Comment (empirephoenix) [reply src/i2c.rs:52]: Here comes the fun part, the idf frameworks accepts 23 but the documentation for the hardware says only up to 22 is allowed. In fact some larger timeouts seem to work, but cause glitch behavior in other parts of the hardware, so the idea here was to ensure only valid values according to the documentation can be set.

File: src/i2c.rs (hunk 4 above)
Comment (empirephoenix) [reply src/i2c.rs:48]: There seems to exist I2C_LL_MAX_TIMEOUT, at least the arduino hal seems to use that, but it seems to not be included in the current bindings, how would I add this?

File: src/i2c.rs (hunk 1 above)
Comment (empirephoenix) [reply src/i2c.rs:40]: Ah yes, removed variable, it is u128 tho

**Comment by ivmarkov** [comment]
Thanks!

**Pull request merged**